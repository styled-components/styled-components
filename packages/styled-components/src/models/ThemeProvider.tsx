import React, { useContext, useMemo } from 'react';
import styledError from '../utils/error';
import isFunction from '../utils/isFunction';

export type Theme = {
  [key: string]: any;
};

type ThemeFn = (outerTheme?: Theme) => Theme;
type ThemeArgument = Theme | ThemeFn;

type Props = {
  children?: React.ReactChild;
  theme: ThemeArgument;
};

export const ThemeContext = React.createContext<Theme | undefined>(undefined);

export const ThemeConsumer = ThemeContext.Consumer;

function mergeTheme(theme: ThemeArgument, outerTheme?: Theme): Theme {
  if (!theme) {
    throw styledError(14);
  }

  if (isFunction(theme)) {
    const themeFn = theme as ThemeFn;
    const mergedTheme = themeFn(outerTheme);

    if (
      process.env.NODE_ENV !== 'production' &&
      (mergedTheme === null || Array.isArray(mergedTheme) || typeof mergedTheme !== 'object')
    ) {
      throw styledError(7);
    }

    return mergedTheme;
  }

  if (Array.isArray(theme) || typeof theme !== 'object') {
    throw styledError(8);
  }

  return outerTheme ? { ...outerTheme, ...theme } : theme;
}

/**
 * Provide a theme to an entire react component tree via context
 */
export default function ThemeProvider(props: Props): JSX.Element | null {
  const outerTheme = useContext(ThemeContext);
  const themeContext = useMemo(() => mergeTheme(props.theme, outerTheme), [
    props.theme,
    outerTheme,
  ]);

  if (!props.children) {
    return null;
  }

  return <ThemeContext.Provider value={themeContext}>{props.children}</ThemeContext.Provider>;
}
