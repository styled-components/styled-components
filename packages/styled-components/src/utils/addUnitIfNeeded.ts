import unitless from '@emotion/unitless';

// Taken from https://github.com/facebook/react/blob/b87aabdfe1b7461e7331abb3601d9e6bb27544bc/packages/react-dom/src/shared/dangerousStyleValue.js
export default function addUnitIfNeeded(name: string, value: any) {
  // https://github.com/amilajack/eslint-plugin-flowtype-errors/issues/133
  if (value == null || typeof value === 'boolean' || value === '') {
    return '';
  }

  if (typeof value === 'number' && value !== 0 && !(name in unitless)) {
    return `${value}px`; // Presumes implicit 'px' suffix for unitless numbers
  }

  return String(value).trim();
}
