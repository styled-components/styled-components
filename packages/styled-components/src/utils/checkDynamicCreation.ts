import { useRef } from 'react';

const invalidHookCallRe = /invalid hook call/i;
const seen = new Set();

export const checkDynamicCreation = (displayName: string, componentId?: string) => {
  if (process.env.NODE_ENV !== 'production') {
    const parsedIdString = componentId ? ` with the id of "${componentId}"` : '';
    const message =
      `The component ${displayName}${parsedIdString} has been created dynamically.\n` +
      "You may see this warning because you've called styled inside another component.\n" +
      'To resolve this only create new StyledComponents outside of any render method and function component.';

    // If a hook is called outside of a component:
    // React 17 and earlier throw an error
    // React 18 and above use console.error

    const originalConsoleError = console.error
    try {
      let didNotCallInvalidHook = true
      console.error = (consoleErrorMessage, ...consoleErrorArgs) => {
      // The error here is expected, since we're expecting anything that uses `checkDynamicCreation` to
      // be called outside of a React component.
        if (invalidHookCallRe.test(consoleErrorMessage)) {
          didNotCallInvalidHook = false
          // This shouldn't happen, but resets `warningSeen` if we had this error happen intermittently
          seen.delete(message);
        } else {
          originalConsoleError(consoleErrorMessage, ...consoleErrorArgs);
        }
      }
      // We purposefully call `useRef` outside of a component and expect it to throw
      // If it doesn't, then we're inside another component.
      // eslint-disable-next-line react-hooks/rules-of-hooks
      useRef();

      if (didNotCallInvalidHook && !seen.has(message)) {
        // eslint-disable-next-line no-console
        console.warn(message);
        seen.add(message);
      }
    } catch (error) {
      // The error here is expected, since we're expecting anything that uses `checkDynamicCreation` to
      // be called outside of a React component.
      if (invalidHookCallRe.test(error.message)) {
        // This shouldn't happen, but resets `warningSeen` if we had this error happen intermittently
        seen.delete(message);
      }
    } finally {
      console.error = originalConsoleError;
    }
  }
};
