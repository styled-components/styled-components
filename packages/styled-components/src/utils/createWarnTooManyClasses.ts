import { ExtensibleObject } from '../types';

export const LIMIT = 200;

export default (displayName: string, componentId: string) => {
  let generatedClasses: ExtensibleObject = {};
  let warningSeen = false;

  return (className: string) => {
    if (!warningSeen) {
      generatedClasses[className] = true;
      if (Object.keys(generatedClasses).length >= LIMIT) {
        // Unable to find latestRule in test environment.

        /* eslint-disable no-console, prefer-template */
        const parsedIdString = componentId ? ` with the id of "${componentId}"` : '';

        console.warn(
          `Over ${LIMIT} classes were generated for component ${displayName}${parsedIdString}.\n` +
            'Consider using the attrs method, together with a style object for frequently changed styles.\n' +
            'Example:\n' +
            '  const Component = styled.div.attrs(props => ({\n' +
            '    style: {\n' +
            '      background: props.background,\n' +
            '    },\n' +
            '  }))`width: 100%;`\n\n' +
            '  <Component />'
        );
        warningSeen = true;
        generatedClasses = {};
      }
    }
  };
};
