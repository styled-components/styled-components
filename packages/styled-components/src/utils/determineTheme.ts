import { ExtensibleObject } from '../types';
import { EMPTY_OBJECT } from './empties';

export default function determineTheme(
  props: ExtensibleObject,
  providedTheme: any,
  defaultProps: any = EMPTY_OBJECT
) {
  return (props.theme !== defaultProps.theme && props.theme) || providedTheme || defaultProps.theme;
}
