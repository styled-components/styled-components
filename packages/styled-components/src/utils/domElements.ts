// Thanks to ReactDOMFactories for this handy list!

export default [
  'a',
  'abbr',
  'address',
  'area',
  'article',
  'aside',
  'audio',
  'b',
  'base',
  'bdi',
  'bdo',
  'big',
  'blockquote',
  'body',
  'br',
  'button',
  'canvas',
  'caption',
  'cite',
  'code',
  'col',
  'colgroup',
  'data',
  'datalist',
  'dd',
  'del',
  'details',
  'dfn',
  'dialog',
  'div',
  'dl',
  'dt',
  'em',
  'embed',
  'fieldset',
  'figcaption',
  'figure',
  'footer',
  'form',
  'h1',
  'h2',
  'h3',
  'h4',
  'h5',
  'h6',
  'head',
  'header',
  'hgroup',
  'hr',
  'html',
  'i',
  'iframe',
  'img',
  'input',
  'ins',
  'kbd',
  'keygen',
  'label',
  'legend',
  'li',
  'link',
  'main',
  'map',
  'mark',
  'menu',
  'menuitem',
  'meta',
  'meter',
  'nav',
  'noscript',
  'object',
  'ol',
  'optgroup',
  'option',
  'output',
  'p',
  'param',
  'picture',
  'pre',
  'progress',
  'q',
  'rp',
  'rt',
  'ruby',
  's',
  'samp',
  'script',
  'section',
  'select',
  'small',
  'source',
  'span',
  'strong',
  'style',
  'sub',
  'summary',
  'sup',
  'table',
  'tbody',
  'td',
  'textarea',
  'tfoot',
  'th',
  'thead',
  'time',
  'title',
  'tr',
  'track',
  'u',
  'ul',
  'var',
  'video',
  'wbr', // SVG
  'circle',
  'clipPath',
  'defs',
  'ellipse',
  'foreignObject',
  'g',
  'image',
  'line',
  'linearGradient',
  'marker',
  'mask',
  'path',
  'pattern',
  'polygon',
  'polyline',
  'radialGradient',
  'rect',
  'stop',
  'svg',
  'text',
  'tspan',
] as const;
