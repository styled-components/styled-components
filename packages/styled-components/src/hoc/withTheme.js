// @flow
import React, { useContext, type AbstractComponent } from 'react';
import hoistStatics from 'hoist-non-react-statics';
import { ThemeContext } from '../models/ThemeProvider';
import determineTheme from '../utils/determineTheme';
import getComponentName from '../utils/getComponentName';

// NOTE: this would be the correct signature:
// export default <Config: { theme?: any }, Instance>(
//  Component: AbstractComponent<Config, Instance>
// ): AbstractComponent<$Diff<Config, { theme?: any }> & { theme?: any }, Instance>
//
// but the old build system tooling doesn't support the syntax

export default (Component: AbstractComponent<*, *>) => {
  // $FlowFixMe This should be React.forwardRef<Config, Instance>
  const WithTheme = React.forwardRef((props, ref) => {
    const theme = useContext(ThemeContext);
    // $FlowFixMe defaultProps isn't declared so it can be inferrable
    const { defaultProps } = Component;
    const themeProp = determineTheme(props, theme, defaultProps);

    if (process.env.NODE_ENV !== 'production' && themeProp === undefined) {
      // eslint-disable-next-line no-console
      console.warn(
        `[withTheme] You are not using a ThemeProvider nor passing a theme prop or a theme in defaultProps in component class "${getComponentName(
          Component
        )}"`
      );
    }

    return <Component {...props} theme={themeProp} ref={ref} />;
  });

  hoistStatics(WithTheme, Component);

  WithTheme.displayName = `WithTheme(${getComponentName(Component)})`;
  
  // The underlying component is expected to get a theme as a prop.
  // However, if the underlying component is folded up (see StyledComponent.js),
  // we cannot tell that it needs a theme as a prop without a mark.
  // So we mark WithTheme with withTheme flag.
  WithTheme.withTheme = true;

  return WithTheme;
};
