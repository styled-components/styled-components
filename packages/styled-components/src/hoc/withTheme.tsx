import React, { ComponentType, useContext } from 'react';
import { ThemeContext } from '../models/ThemeProvider';
import determineTheme from '../utils/determineTheme';
import getComponentName from '../utils/getComponentName';
import hoist from '../utils/hoist';

export default function withTheme(Component: ComponentType<any>) {
  const WithTheme = React.forwardRef((props, ref) => {
    const theme = useContext(ThemeContext);
    const themeProp = determineTheme(props, theme, Component.defaultProps);

    if (process.env.NODE_ENV !== 'production' && themeProp === undefined) {
      // eslint-disable-next-line no-console
      console.warn(
        `[withTheme] You are not using a ThemeProvider nor passing a theme prop or a theme in defaultProps in component class "${getComponentName(
          Component
        )}"`
      );
    }

    return <Component {...props} theme={themeProp} ref={ref} />;
  });

  hoist(WithTheme, Component);

  WithTheme.displayName = `WithTheme(${getComponentName(Component)})`;

  return WithTheme;
}
