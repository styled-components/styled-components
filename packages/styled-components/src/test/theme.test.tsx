
import React, { Component } from "react";
import { renderIntoDocument } from "react-dom/test-utils";
import TestRenderer from "react-test-renderer";
import withTheme from "../hoc/withTheme";
import ThemeProvider from "../models/ThemeProvider";
import { getRenderedCSS, resetStyled } from "./utils";

let styled: ReturnType<typeof resetStyled>;

describe('theming', () => {
  beforeEach(() => {
    jest.spyOn(console, 'error').mockImplementation(() => {});
    styled = resetStyled();
  });

  it('should inject props.theme into a styled component', () => {
    const Comp = styled.div`
      color: ${props => props.theme.color};
    `;
    const theme = { color: 'black' };
    TestRenderer.create(<ThemeProvider theme={theme}>
        <Comp />
      </ThemeProvider>);
    expect(getRenderedCSS()).toMatchInlineSnapshot(`
      ".b {
        color: black;
      }"
    `);
  });

  it('should inject props.theme into a styled component multiple levels deep', () => {
    const Comp = styled.div`
      color: ${props => props.theme.color};
    `;
    const theme = { color: 'black' };
    TestRenderer.create(<ThemeProvider theme={theme}>
        <div>
          <div>
            <Comp />
          </div>
        </div>
      </ThemeProvider>);
    expect(getRenderedCSS()).toMatchInlineSnapshot(`
      ".b {
        color: black;
      }"
    `);
  });

  it('should properly allow a component to fallback to its default props when a theme is not provided', () => {
    const Comp1 = styled.div`
      color: ${props => props.theme.test.color};
    `;

    Comp1.defaultProps = {
      theme: {
        test: {
          color: 'purple'
        }
      }
    };
    TestRenderer.create(<div>
        <Comp1 />
      </div>);
    expect(getRenderedCSS()).toMatchInlineSnapshot(`
      ".b {
        color: purple;
      }"
    `);
  });

  // https://github.com/styled-components/styled-components/issues/344
  it('should use ThemeProvider theme instead of defaultProps theme', () => {
    const Comp1 = styled.div`
      color: ${props => props.theme.test.color};
    `;

    Comp1.defaultProps = {
      theme: {
        test: {
          color: 'purple'
        }
      }
    };
    const theme = { test: { color: 'green' } };

    TestRenderer.create(<ThemeProvider theme={theme}>
        <Comp1 />
      </ThemeProvider>);
    expect(getRenderedCSS()).toMatchInlineSnapshot(`
      ".b {
        color: green;
      }"
    `);
  });

  it('should properly allow a component to override the theme with a prop even if it is equal to defaultProps theme', () => {
    const Comp1 = styled.div`
      color: ${props => props.theme.test.color};
    `;

    Comp1.defaultProps = {
      theme: {
        test: {
          color: 'purple'
        }
      }
    };
    const theme = { test: { color: 'green' } };

    TestRenderer.create(<ThemeProvider theme={theme}>
        <Comp1 theme={{ test: { color: 'purple' } }} />
      </ThemeProvider>);
    expect(getRenderedCSS()).toMatchInlineSnapshot(`
      ".b {
        color: purple;
      }"
    `);
  });

  it('should properly allow a component to override the theme with a prop', () => {
    const Comp = styled.div`
      color: ${props => props.theme.color};
    `;

    const theme = {
      color: 'purple'
    };

    TestRenderer.create(<div>
        <ThemeProvider theme={theme}>
          <Comp theme={{ color: 'red' }} />
        </ThemeProvider>
      </div>);
    expect(getRenderedCSS()).toMatchInlineSnapshot(`
      ".b {
        color: red;
      }"
    `);
  });

  it('should only inject props.theme into styled components within its child component tree', () => {
    const Comp1 = styled.div`
      color: ${props => props.theme.color || 'red'};
    `;
    const Comp2 = styled.div`
      color: ${props => props.theme.color || 'red'};
    `;

    const theme = { color: 'black' };
    TestRenderer.create(<div>
        <ThemeProvider theme={theme}>
          <div>
            <Comp1 />
          </div>
        </ThemeProvider>
        <Comp2 />
      </div>);
    expect(getRenderedCSS()).toMatchInlineSnapshot(`
      ".c {
        color: black;
      }
      .d {
        color: red;
      }"
    `);
  });

  it('should inject props.theme into all styled components within the child component tree', () => {
    const Comp1 = styled.div`
      color: ${props => props.theme.color};
    `;
    const Comp2 = styled.div`
      background: ${props => props.theme.color};
    `;
    const theme = { color: 'black' };
    TestRenderer.create(<ThemeProvider theme={theme}>
        <div>
          <div>
            <Comp1 />
          </div>
          <Comp2 />
        </div>
      </ThemeProvider>);
    expect(getRenderedCSS()).toMatchInlineSnapshot(`
      ".c {
        color: black;
      }
      .d {
        background: black;
      }"
    `);
  });

  it('should inject new CSS when the theme changes', () => {
    const Comp = styled.div`
      color: ${props => props.theme.color};
    `;
    const originalTheme = { color: 'black' };
    const newTheme = { color: 'blue' };
    let theme = originalTheme;
    // Force render the component
    const renderComp = () => {
      TestRenderer.create(<ThemeProvider theme={theme}>
          <Comp />
        </ThemeProvider>);
    };
    renderComp();
    expect(getRenderedCSS()).toMatchInlineSnapshot(`
      ".b {
        color: black;
      }"
    `);
    // Change the theme
    theme = newTheme;
    renderComp();
    expect(getRenderedCSS()).toMatchInlineSnapshot(`
      ".b {
        color: black;
      }
      .c {
        color: blue;
      }"
    `);
  });

  it('should properly render with the same theme from default props on re-render', () => {
    const Comp1 = styled.div`
      color: ${props => props.theme.color};
    `;

    Comp1.defaultProps = {
      theme: {
        color: 'purple'
      }
    };

    const jsx = <Comp1 />;

    const wrapper = TestRenderer.create(jsx);
    expect(getRenderedCSS()).toMatchInlineSnapshot(`
      ".b {
        color: purple;
      }"
    `);

    wrapper.update(jsx);
    expect(getRenderedCSS()).toMatchInlineSnapshot(`
      ".b {
        color: purple;
      }"
    `);
  });

  it('should properly update style if theme is changed', () => {
    const Comp1 = styled.div`
      color: ${props => props.theme.color};
    `;

    const wrapper = TestRenderer.create(<ThemeProvider theme={{
      color: 'purple'
    }}>
        <Comp1 />
      </ThemeProvider>);
    expect(getRenderedCSS()).toMatchInlineSnapshot(`
      ".b {
        color: purple;
      }"
    `);

    wrapper.update(<ThemeProvider theme={{
      color: 'pink'
    }}>
        <Comp1 />
      </ThemeProvider>);
    expect(getRenderedCSS()).toMatchInlineSnapshot(`
      ".b {
        color: purple;
      }
      .c {
        color: pink;
      }"
    `);
  });

  it('should properly update style if props used in styles is changed', () => {
    const Comp1 = styled.div`
      color: ${props => props.theme.color};
      z-index: ${props => props.zIndex};
    `;

    Comp1.defaultProps = {
      zIndex: 0
    };

    const wrapper = TestRenderer.create(<ThemeProvider theme={{
      color: 'purple'
    }}>
        <Comp1 />
      </ThemeProvider>);

    expect(getRenderedCSS()).toMatchInlineSnapshot(`
      ".b {
        color: purple;
        z-index: 0;
      }"
    `);

    wrapper.update(<ThemeProvider theme={{
      color: 'pink'
    }}>
        <Comp1 />
      </ThemeProvider>);

    expect(getRenderedCSS()).toMatchInlineSnapshot(`
      ".b {
        color: purple;
        z-index: 0;
      }
      .c {
        color: pink;
        z-index: 0;
      }"
    `);

    Comp1.defaultProps.zIndex = 1;
    wrapper.update(<ThemeProvider theme={{
      color: 'pink'
    }}>
        <Comp1 />
      </ThemeProvider>);
    expect(getRenderedCSS()).toMatchInlineSnapshot(`
      ".b {
        color: purple;
        z-index: 0;
      }
      .c {
        color: pink;
        z-index: 0;
      }
      .d {
        color: pink;
        z-index: 1;
      }"
    `);
  });

  it('should change the classnames when the theme changes', () => {
    const Comp = styled.div`
      color: ${props => props.theme.color};
    `;

    const originalTheme = { color: 'black' };
    const newTheme = { color: 'blue' };

    const Theme = ({
      theme
    }) => <ThemeProvider theme={theme}>
        <Comp someProps={theme} />
      </ThemeProvider>;

    const wrapper = TestRenderer.create(<Theme theme={originalTheme} />);

    expect(getRenderedCSS()).toMatchInlineSnapshot(`
      ".b {
        color: black;
      }"
    `);
    expect(wrapper.root.findByType('div').props.className).toBe('sc-a b');

    // Change theme
    wrapper.update(Theme({ theme: newTheme }));

    expect(getRenderedCSS()).toMatchInlineSnapshot(`
      ".b {
        color: black;
      }
      .c {
        color: blue;
      }"
    `);

    expect(wrapper.root.findByType('div').props.className).toBe('sc-a c');
  });

  it('should inject props.theme into a component that uses withTheme hoc', () => {
    const originalTheme = { color: 'black' };

    const MyDiv = ({
      theme
    }) => <div>{theme.color}</div>;
    const MyDivWithTheme = withTheme(MyDiv);

    const wrapper = TestRenderer.create(<ThemeProvider theme={originalTheme}>
        <MyDivWithTheme />
      </ThemeProvider>);

    expect(wrapper.root.findByType('div').props.children).toBe('black');
  });

  it('should properly update theme prop on hoc component when theme is changed', () => {
    const MyDiv = ({
      theme
    }) => <div>{theme.color}</div>;
    const MyDivWithTheme = withTheme(MyDiv);

    const originalTheme = { color: 'black' };
    const newTheme = { color: 'blue' };

    const Theme = ({
      theme
    }) => <ThemeProvider theme={theme}>
        <MyDivWithTheme />
      </ThemeProvider>;

    const wrapper = TestRenderer.create(<Theme key="a" theme={originalTheme} />);
    expect(wrapper.root.findByType('div').props.children).toBe('black');

    // Change theme
    wrapper.update(<Theme key="a" theme={newTheme} />);

    expect(wrapper.root.findByType('div').props.children).toBe('blue');
  });

  // https://github.com/styled-components/styled-components/issues/445
  it('should use ThemeProvider theme instead of defaultProps theme after initial render', () => {
    const Text = styled.div`
      color: ${props => props.theme.color};
    `;

    Text.defaultProps = {
      theme: {
        color: 'purple'
      }
    };

    const Theme = props => <ThemeProvider theme={{ color: 'green' }}>
        <Text {...props} />
      </ThemeProvider>;

    const wrapper = TestRenderer.create(<Theme key="a" prop="foo" />);
    expect(getRenderedCSS()).toMatchInlineSnapshot(`
      ".b {
        color: green;
      }"
    `);

    wrapper.update(<Theme key="a" prop="bar" />);
    expect(getRenderedCSS()).toMatchInlineSnapshot(`
      ".b {
        color: green;
      }"
    `);
  });

  // https://github.com/styled-components/styled-components/issues/596
  it('should hoist static properties when using withTheme', () => {
    class MyComponent extends Component<any, any> {

      static myStaticProperty = true;
    }

    const MyComponentWithTheme = withTheme(MyComponent);

    expect(MyComponentWithTheme.myStaticProperty).toBe(true);
  });

  it('should only pass the theme prop', () => {
    class Comp extends Component<any, any> {

      render() {
        return <div />;
      }
    }

    const CompWithTheme = withTheme(Comp);

    const wrapper = TestRenderer.create(<ThemeProvider theme={{}}>
        <CompWithTheme />
      </ThemeProvider>);

    const inner = wrapper.root.findByType(Comp);

    expect(Object.keys(inner.props).length).toEqual(1);
    expect(inner.props).toEqual({ theme: {} });
  });

  it('should forward refs', () => {
    class Comp extends Component<any, any> {

      render() {
        return <div {...this.props} />;
      }
    }

    const CompWithTheme = withTheme(Comp);
    const ref = React.createRef();

    renderIntoDocument(<ThemeProvider theme={{}}>
        <CompWithTheme ref={ref} />
      </ThemeProvider>);

    expect(ref.current).toBeInstanceOf(Comp);
  });

  // https://github.com/styled-components/styled-components/issues/1130
  it('should not break without a ThemeProvider if it has a defaultTheme', () => {
    const MyDiv = ({
      theme
    }) => <div>{theme.color}</div>;
    const MyDivWithTheme = withTheme(MyDiv);
    const theme = { color: 'red' };
    const newTheme = { color: 'blue' };

    // eslint-disable-next-line no-console
    const consoleWarn = console.warn;

    jest.spyOn(console, 'warn').mockImplementation(msg => !msg.includes('You are not using a ThemeProvider') ? consoleWarn(msg) : null);

    MyDivWithTheme.defaultProps = { theme };

    const wrapper = TestRenderer.create(<MyDivWithTheme />);

    expect(wrapper.root.findByType('div').props.children).toBe('red');

    // Change theme
    MyDivWithTheme.defaultProps = { theme: newTheme };

    // Change theme
    wrapper.update(<MyDivWithTheme />);

    expect(wrapper.root.findByType('div').props.children).toBe('blue');
  });

  // https://github.com/styled-components/styled-components/issues/1776
  it('should allow module objects to be passed as themes', () => {
    const theme = {
      borderRadius: '2px',
      palette: {
        black: '#000',
        white: '#fff',
        // Flow has limited support for Symbols and computed properties;
        // see <https://github.com/facebook/flow/issues/3258>.
        // $FlowFixMe
        [Symbol.toStringTag]: 'Module'
      },
      // Flow has limited support for Symbols and computed properties;
      // see <https://github.com/facebook/flow/issues/3258>.
      // $FlowFixMe
      [Symbol.toStringTag]: 'Module'
    };

    const Comp1 = styled.div`
      background-color: ${({
      theme
    }) => theme.palette.white};
      color: ${({
      theme
    }) => theme.palette.black};
    `;

    expect(() => {
      TestRenderer.create(<ThemeProvider theme={theme}>
          <Comp1 />
        </ThemeProvider>);
    }).not.toThrow('plain object');

    expect(getRenderedCSS()).toMatchInlineSnapshot(`
      ".b {
        background-color: #fff;
        color: #000;
      }"
    `);
  });

  it('should allow other complex objects to be passed as themes', () => {
    class Theme {

      borderRadius: string;

      constructor(borderRadius) {
        this.borderRadius = borderRadius;
      }
    }

    const theme = new Theme('2px');

    const Comp1 = styled.div`
      border-radius: ${({
      theme
    }) => theme.borderRadius};
    `;

    TestRenderer.create(<ThemeProvider theme={theme}>
        <Comp1 />
      </ThemeProvider>);

    expect(getRenderedCSS()).toMatchInlineSnapshot(`
      ".b {
        border-radius: 2px;
      }"
    `);
  });

  it('should not allow the theme to be null', () => {
    expect(() => {
      // HACK: work around the problem without changing the snapshots
      // these tests need to be changed to use error boundaries instead
      const mock = jest.spyOn(console, 'error').mockImplementation(() => {});
      TestRenderer.create( // $FlowInvalidInputTest
      <ThemeProvider theme={null}>
          <div />
        </ThemeProvider>);
      expect(mock).toHaveBeenCalledTimes(1);
    }).toThrowErrorMatchingSnapshot();
  });

  it('should not allow the theme to be an array', () => {
    expect(() => {
      // HACK: work around the problem without changing the snapshots
      // these tests need to be changed to use error boundaries instead
      const mock = jest.spyOn(console, 'error').mockImplementation(() => {});
      TestRenderer.create( // $FlowInvalidInputTest
      <ThemeProvider theme={['a', 'b', 'c']}>
          <div />
        </ThemeProvider>);
      expect(mock).toHaveBeenCalledTimes(1);
    }).toThrowErrorMatchingSnapshot();
  });

  it('should not allow the theme to be a non-object', () => {
    expect(() => {
      // HACK: work around the problem without changing the snapshots
      // these tests need to be changed to use error boundaries instead
      const mock = jest.spyOn(console, 'error').mockImplementation(() => {});
      TestRenderer.create( // $FlowInvalidInputTest
      <ThemeProvider theme={42}>
          <div />
        </ThemeProvider>);
      expect(mock).toHaveBeenCalledTimes(1);
    }).toThrowErrorMatchingSnapshot();
  });
});