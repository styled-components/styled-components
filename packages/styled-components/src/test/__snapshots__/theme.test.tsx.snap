// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`theming should not allow the theme to be a non-object 1`] = `"ThemeProvider: Please make your \\"theme\\" prop an object."`;

exports[`theming should not allow the theme to be an array 1`] = `"ThemeProvider: Please make your \\"theme\\" prop an object."`;

exports[`theming should not allow the theme to be null 1`] = `"ThemeProvider: \\"theme\\" prop is required."`;
